File: 2024/6.py
import util


def part1(input: str) -> int:
    lines = input.splitlines()
    start = None
    for r in range(len(lines)):
        for c in range(len(lines[0])):
            if lines[r][c] == "^":
                start = (r, c)

    pos = set()
    r, c = start
    dr, dc = -1, 0
    while True:
        pos.add((r, c))
        try:
            if lines[r + dr][c + dc] == "#":
                dr, dc = dc, -dr
            else:
                r += dr
                c += dc
        except IndexError:
            return len(pos)


def has_loop(grid, h, w, start) -> bool:
    pos = set()
    r, c = start
    dr, dc = -1, 0
    while True:
        if (r, c, dr, dc) in pos:
            return True
        pos.add((r, c, dr, dc))

        match grid.get((r + dr, c + dc)):
            case None:
                return False
            case "#":
                dr, dc = dc, -dr
            case _:
                r += dr
                c += dc


def part2(input: str) -> int:
    lines = input.splitlines()
    h, w = len(lines), len(lines[0])
    grid = {}
    start = None
    for r in range(h):
        for c in range(w):
            if lines[r][c] == "^":
                start = (r, c)
            grid[(r, c)] = lines[r][c]

    pos = set()
    r, c = start
    dr, dc = -1, 0
    while True:
        pos.add((r, c))
        try:
            if lines[r + dr][c + dc] == "#":
                dr, dc = dc, -dr
            else:
                r += dr
                c += dc
        except IndexError:
            break

    x = 0
    for block in pos:
        if block == start:
            continue

        m, grid[block] = grid[block], "#"
        if has_loop(grid, h, w, start):
            x += 1
        grid[block] = m

    return x


if __name__ == "__main__":
    util.run(6, part1, part2)

File: 2025/6.ts
import { run } from "@util";

async function part1(input: string): Promise<number> {
  const lines = input.split("\n");
  const ops = new Map<number, string>();
  const totals = new Map<number, number>();

  const opsLine = lines[lines.length - 1].split(/\s+/);
  for (let i = 0; i < opsLine.length; i++) {
    ops.set(i, opsLine[i]);
  }

  for (let i = lines.length - 2; i >= 0; i--) {
    const nums = lines[i]
      .split(/\s+/)
      .filter((v) => v !== "")
      .map(Number);

    for (let j = 0; j < nums.length; j++) {
      const fn = ops.get(j);
      let curr;
      switch (fn) {
        case "+":
          curr = totals.get(j) || 0;
          totals.set(j, curr + nums[j]);
          break;
        case "*":
          curr = totals.get(j) || 1;
          totals.set(j, curr * nums[j]);
          break;
      }
    }
  }

  return Array.from(totals.values()).reduce((acc, val) => acc + val, 0);
}

async function part2(input: string): Promise<number> {
  const lines = input.split("\n");

  const ops = new Map<number, string>();
  const lastLine = lines[lines.length - 1].split("");
  for (let i = 0; i < lastLine.length; i++) {
    const char = lastLine[i];
    if (char === "*" || char === "+") {
      ops.set(i, char);
    }
  }

  const rawNums: Array<Array<string | undefined>> = [];
  for (let r = 0; r < lines[0].length; r++) {
    const row = [];
    for (let c = 0; c < lines.length - 1; c++) {
      row.push(undefined);
    }
    rawNums.push(row);
  }

  for (let c = lines.length - 2; c >= 0; c--) {
    const line = lines[c];
    for (let r = 0; r < line.length; r++) {
      rawNums[r][c] = line[r];
    }
  }
  const nums: Array<number | undefined> = rawNums
    .map((n) => n.filter(Boolean).join(""))
    .map((n) => (n.match(/^\s+$/) ? undefined : Number(n)));

  let total = 0;
  let currOp = ops.get(0);
  let currTotal = nums[0];

  if (currTotal === undefined) {
    throw new Error("impossible");
  }

  for (let i = 1; i < nums.length; i++) {
    const num = nums[i];
    currOp = ops.get(i) || currOp;

    if (num === undefined) {
      total += currTotal;
      currTotal = ops.get(i + 1) === "*" ? 1 : 0;
      continue;
    }

    switch (currOp) {
      case "+":
        currTotal += num;
        break;
      case "*":
        currTotal *= num;
        break;
    }
  }

  return total + currTotal;
}

run(6, 2025, part1, part2);
